plugins/module_utils/network/sonic/config/sflow/sflow.py

```diff
@@ -0,0 +1,522 @@
+            requests.append({"path": "data/openconfig-sampling-sflow:sampling/sflow/config/agent", "method": "DELETE"})
+        if "sampling_rate" in to_delete:
+            requests.append({"path": "data/openconfig-sampling-sflow:sampling/sflow/config/sampling-rate", "method": "DELETE"})
+        if "collectors" in to_delete:
+            have_collectors_dict = {(collector["address"], collector["network_instance"], collector["port"]): collector for collector in have["collectors"]}
+            to_delete_collectors_dict = {(collector["address"], collector["network_instance"], collector["port"]): collector
+                                         for collector in to_delete["collectors"]}
+            if have_collectors_dict.keys() == to_delete_collectors_dict.keys():
+                requests.append({"path": "data/openconfig-sampling-sflow:sampling/sflow/collectors", "method": "DELETE"})
+            else:
+                for collector in to_delete["collectors"]:
+                    requests.append({"path": "data/openconfig-sampling-sflow:sampling/sflow/collectors/collector=" +
+                                    collector["address"] + "," + str(collector["port"]) + "," +
+                                    collector["network_instance"], "method": "DELETE"})
+        if "interfaces" in to_delete:
+            for interface in to_delete["interfaces"]:
+                requests.append({"path": "data/openconfig-sampling-sflow:sampling/sflow/interfaces/interface=" + interface["name"], "method": "DELETE"})
+        return requests
+
+    def fill_defaults(self, config):
+        '''modifies the given original config object to add sflow default values that are missing. returns the config for chaining purposes'''
+        if "enabled" not in config:
+            config["enabled"] = False
+        return config
+
+    def find_substitution_deletes(self, remove_diff, introduced_diff):
+        '''specifically for overridden and replaced states, finds and builds collection of which config settings need to be deleted and without anything that is
+        getting replaced with new values. `get_diff` will return collection of both things that need to be deleted and things that will have new values.'''
+        result = {}
+        if "agent" in remove_diff and "agent" not in introduced_diff:
+            result["agent"] = remove_diff["agent"]
+        if "enabled" in remove_diff and "enabled" not in introduced_diff:
+            result["enabled"] = remove_diff["enabled"]
+        if "polling_interval" in remove_diff and "polling_interval" not in introduced_diff:
+            result["polling_interval"] = remove_diff["polling_interval"]
+        if "sampling_rate" in remove_diff and "sampling_rate" not in introduced_diff:
+            result["sampling_rate"] = remove_diff["sampling_rate"]
+        if "collectors" in remove_diff:
+            result["collectors"] = remove_diff["collectors"]
+        if "interfaces" in remove_diff:
```

Comment:
To filter out deletes for interfaces that are being changed ("substitution" deletes) so they don't get deleted unnecessarily, a loop in this block would be helpful.

For example:

           
           for interface_r in remove_diff['interfaces']:
               substituted = False
               rname = interface_r['name']
               for interface_i in introduced_diff['interfaces']:
                   if interface_i['name'] == rname:
                     substituted = True;
                     break
      
               if not substituted:
                   result["interfaces"].append(interface_r)


The block shown above would replace the current instruction (below) in this "if" block which puts the entire "remove_diff["interfaces"] list into the result, even if some of the interfaces have substitutions.

Reply:
sflow merge uses patch requests so if there's a setting currently configured that's not in want, it isn't going to be deleted without overridden doing extra steps and call to delete. This sounds good for reducing calls, but I think just checking for whether name is in both to decided if deletion should be done will miss some settings within interfaces and should check settings as well.

Reply:
I agree with the considerations you have presented here.

To preserve the intent of the "remove_substitution_deletes" operation here, the loop shown above could be implemented with the modified algorithm shown above to handle the "substituted" case in which the interface is present in both lists.

The point of this is to enable removal of the prep_merge_diff,  build_ref_dict, and find_item_key functions by adding within this function a smaller amount of code than the sum of those three functions .

Reply:
After thinking about this more, I would like to bring up another option that  you can consider. I am not favoring this over the current implementation, but want to bring it up to clarify the "big picture" for implementation of "overridden" state in general.)

Logically, "overidden" state implementation is the following:

- If no changes exist between the current "have" configuration and the playbook "want" configuration, take no further action. This situation can easily be recognized by determining if "get_diff(want, have)" and "get_diff(have, want)" are both empty.
- If any changes of either type are present, delete all configuration in "have", then merge all configuration in "want".

Anything in the implementation that goes beyond the description above is just optimization. It can be desirable for performance reasons in many cases, but we need to balance it against code complexity.

In this case, to simplify the code, it would be acceptable to do something like this:

Replace the current code from lines 177-197, starting at " remove_diff = self.find_substitution_deletes..." with the following:

        if remove_diff is not None and introduced_diff is not None:
            commands, requests = self._state_deleted(have, have)
        commandsTwo, requestsTwo = self._state_merged(want, [], "overridden")  << See related comment for parm 3
  
        commands.extend(commandsTwo)
        requests.extend(requestsTwo)

        return commands, requests